segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/UtilsWrapper.java

```diff
@@ -0,0 +1,287 @@
+
+        return chunkedSegmentStorage.getStreamSegmentInfo(segmentName, timeout)
+                .thenComposeAsync(info -> {
+                    val bytesRemaining = new AtomicLong(info.getLength() - info.getStartOffset());
+                    val offsetToRead = new AtomicLong(info.getStartOffset());
+                    return chunkedSegmentStorage.openRead(segmentName)
+                            .thenComposeAsync(handle -> Futures.loop(
+                                    () -> bytesRemaining.get() > 0,
+                                    () -> chunkedSegmentStorage.read(handle,
+                                                    offsetToRead.get(),
+                                                    buffer,
+                                                    0,
+                                                    Math.toIntExact(Math.min(bytesRemaining.get(), buffer.length)),
+                                                    timeout)
+                                            .thenComposeAsync(bytesRead -> {
+                                                bytesRemaining.addAndGet(-bytesRead);
+                                                offsetToRead.addAndGet(bytesRead);
+                                                try {
+                                                    outputStream.write(buffer, 0, bytesRead);
+                                                    return completedFuture(null);
+                                                } catch (IOException e) {
+                                                    return CompletableFuture.failedFuture(e);
+                                                }
+                                            }, chunkedSegmentStorage.getExecutor()),
+                                    chunkedSegmentStorage.getExecutor()),
+                            chunkedSegmentStorage.getExecutor());
+                }, chunkedSegmentStorage.getExecutor());
+
+    }
+
+    /**
+     * Copy the contents of given chunk to provided {@link OutputStream}.
+     *
+     * @param chunkName Name of the chunk.
+     * @param outputStream Instance of {@link OutputStream} to copy to.
+     * @return A CompletableFuture that, when completed, will indicate that the operation completed.
+     *         If the operation failed, it will be completed with the appropriate exception.
+     */
+    public CompletableFuture<Void> copyFromChunk(String chunkName, OutputStream outputStream) {
+        Preconditions.checkNotNull(chunkName, "chunkName");
+        Preconditions.checkNotNull(outputStream, "outputStream");
+
+        byte[] buffer = new byte[bufferSize];
+
+        return chunkedSegmentStorage.getChunkStorage().getInfo(chunkName)
+                .thenComposeAsync(info -> {
+                    val bytesRemaining = new AtomicLong(info.getLength());
+                    val offsetToRead = new AtomicLong(0);
+                    return chunkedSegmentStorage.getChunkStorage().openRead(chunkName)
```

Comment:
They are not identical

chunkedSegmentStorage.getChunkStorage().openRead  (directly reads one individual chunk)
vs
chunkedSegmentStorage.read (just use regular chunkedSegmentStorage method to read from start to finish - the product code does the real work )

Reply:
Problem is it looks same but isn't really same.
Almost every other line is different. Those subtle differences make it hard/annoying to extract such method.

- val offsetToRead = new AtomicLong(0);  Chunk always starts at 0 where as segment always starts at startOffset
- chunkedSegmentStorage.getChunkStorage().getInfo(chunkName) vs chunkedSegmentStorage.getStreamSegmentInfo(segmentName, timeout) two different calls and types. Two different fields are queried. 
- although actual read calls look similar they have different signature (SegmentHandle and ChunkHandle)

The resulting method after extracting common functionality will have to take 3/4 different lambdas and needs 2/3 template parameters. 
Assuming get past compile issues and somehow extract such a method - it will be completely cryptic and opaque. 
Its a lots of hassle and ugly code without substantial benefits.